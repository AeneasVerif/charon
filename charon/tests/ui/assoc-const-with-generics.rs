struct V<const N: usize, T> {
    x: [T; N],
}

impl<const N: usize, T> V<N, T> {
    const LEN: usize = N; // This has generics <N, T>
}

trait HasLen {
    const LEN: usize;
}

impl<const N: usize> HasLen for [(); N] {
    const LEN: usize = N;
}

impl<const N: usize> HasLen for [bool; N] {
    const LEN: usize = N + 1;
}

/// The assoc const depends on a generic parameter.
struct Wrapper<T>(T);
impl<T: HasLen> HasLen for Wrapper<T> {
    const LEN: usize = T::LEN;
}

pub trait HasDefaultLen<const M: usize> {
    const LEN: usize = M;
    // fn use_array(_a: [(); Self::LEN]) {}
}

impl<const N: usize> HasDefaultLen<N> for [(); N] {}

impl<const N: usize> HasDefaultLen<N> for [bool; N] {
    // Recursive constant because we can
    const LEN: usize = if true {
        N
    } else {
        <Self as HasDefaultLen<N>>::LEN
    };
}

/// The default assoc const depends the `Self` clause: we must be careful when reusing this default
/// value in impls.
trait AlsoHasLen: HasLen {
    const ALSO_LEN: usize = Self::LEN + 1;
}
impl<const N: usize> AlsoHasLen for [(); N] {}
