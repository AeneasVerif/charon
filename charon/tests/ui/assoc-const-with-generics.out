# Final LLBC before serialization:

// Full name: core::marker::MetaSized
#[lang_item("meta_sized")]
pub trait MetaSized<Self>

// Full name: core::marker::Sized
#[lang_item("sized")]
pub trait Sized<Self>
{
    parent_clause0 : [@TraitClause0]: MetaSized<Self>
    non-dyn-compatible
}

// Full name: test_crate::V
struct V<T, const N : usize>
where
    [@TraitClause0]: Sized<T>,
{
  x: Array<T, const N : usize>,
}

fn test_crate::{V<T, const N : usize>[@TraitClause0]}::LEN<T, const N : usize>() -> usize
where
    [@TraitClause0]: Sized<T>,
{
    let @0: usize; // return

    @0 := const (const N : usize)
    return
}

const test_crate::{V<T, const N : usize>[@TraitClause0]}::LEN<T, const N : usize>: usize
where
    [@TraitClause0]: Sized<T>,
 = test_crate::{V<T, const N : usize>[@TraitClause0]}::LEN()

// Full name: test_crate::HasLen
trait HasLen<Self>
{
    parent_clause0 : [@TraitClause0]: MetaSized<Self>
    const LEN : usize
    non-dyn-compatible
}

// Full name: test_crate::{impl HasLen for Array<(), const N : usize>}::LEN
fn {impl HasLen for Array<(), const N : usize>}::LEN<const N : usize>() -> usize
{
    let @0: usize; // return

    @0 := const (const N : usize)
    return
}

// Full name: test_crate::{impl HasLen for Array<(), const N : usize>}::LEN
const {impl HasLen for Array<(), const N : usize>}::LEN<const N : usize>: usize = {impl HasLen for Array<(), const N : usize>}::LEN()

// Full name: test_crate::{impl HasLen for Array<(), const N : usize>}
impl<const N : usize> HasLen for Array<(), const N : usize> {
    parent_clause0 = MetaSized<Array<(), const N : usize>>
    const LEN = {impl HasLen for Array<(), const N : usize>}::LEN<const N : usize>
    non-dyn-compatible
}

// Full name: test_crate::{impl HasLen for Array<bool, const N : usize>}::LEN
fn {impl HasLen for Array<bool, const N : usize>}::LEN<const N : usize>() -> usize
{
    let @0: usize; // return
    let @1: usize; // anonymous local

    storage_live(@1)
    @1 := const (const N : usize) panic.+ const (1 : usize)
    @0 := move (@1)
    return
}

// Full name: test_crate::{impl HasLen for Array<bool, const N : usize>}::LEN
const {impl HasLen for Array<bool, const N : usize>}::LEN<const N : usize>: usize = {impl HasLen for Array<bool, const N : usize>}::LEN()

// Full name: test_crate::{impl HasLen for Array<bool, const N : usize>}
impl<const N : usize> HasLen for Array<bool, const N : usize> {
    parent_clause0 = MetaSized<Array<bool, const N : usize>>
    const LEN = {impl HasLen for Array<bool, const N : usize>}::LEN<const N : usize>
    non-dyn-compatible
}

// Full name: test_crate::Wrapper
struct Wrapper<T>
where
    [@TraitClause0]: Sized<T>,
{
  T,
}

// Full name: test_crate::{impl HasLen for Wrapper<T>[@TraitClause0]}::LEN
fn {impl HasLen for Wrapper<T>[@TraitClause0]}::LEN<T>() -> usize
where
    [@TraitClause0]: Sized<T>,
    [@TraitClause1]: HasLen<T>,
{
    let @0: usize; // return

    @0 := const (@TraitClause1::LEN)
    return
}

// Full name: test_crate::{impl HasLen for Wrapper<T>[@TraitClause0]}::LEN
const {impl HasLen for Wrapper<T>[@TraitClause0]}::LEN<T>: usize
where
    [@TraitClause0]: Sized<T>,
    [@TraitClause1]: HasLen<T>,
 = {impl HasLen for Wrapper<T>[@TraitClause0]}::LEN()

// Full name: test_crate::{impl HasLen for Wrapper<T>[@TraitClause0]}
impl<T> HasLen for Wrapper<T>[@TraitClause0]
where
    [@TraitClause0]: Sized<T>,
    [@TraitClause1]: HasLen<T>,
{
    parent_clause0 = MetaSized<Wrapper<T>[@TraitClause0]>
    const LEN = {impl HasLen for Wrapper<T>[@TraitClause0]}::LEN<T>[@TraitClause0, @TraitClause1]
    non-dyn-compatible
}

// Full name: test_crate::HasDefaultLen
pub trait HasDefaultLen<Self, const M : usize>
{
    parent_clause0 : [@TraitClause0]: MetaSized<Self>
    const LEN : usize
    non-dyn-compatible
}

pub fn test_crate::HasDefaultLen::LEN<Self, const M : usize>() -> usize
where
    [@TraitClause0]: HasDefaultLen<Self, const M : usize>,
{
    let @0: usize; // return

    @0 := const (const M : usize)
    return
}

pub const test_crate::HasDefaultLen::LEN<Self, const M : usize>: usize
where
    [@TraitClause0]: HasDefaultLen<Self, const M : usize>,
 = test_crate::HasDefaultLen::LEN()

// Full name: test_crate::{impl HasDefaultLen<const N : usize> for Array<(), const N : usize>}
impl<const N : usize> HasDefaultLen<const N : usize> for Array<(), const N : usize> {
    parent_clause0 = MetaSized<Array<(), const N : usize>>
    const LEN = test_crate::HasDefaultLen::LEN<Array<(), const N : usize>, const N : usize>[{impl HasDefaultLen<const N : usize> for Array<(), const N : usize>}<const N : usize>]
    non-dyn-compatible
}

// Full name: test_crate::{impl HasDefaultLen<const N : usize> for Array<bool, const N : usize>}::LEN
pub fn {impl HasDefaultLen<const N : usize> for Array<bool, const N : usize>}::LEN<const N : usize>() -> usize
{
    let @0: usize; // return
    let @1: bool; // anonymous local

    storage_live(@1)
    @1 := const (true)
    if move (@1) {
        @0 := const (const N : usize)
    }
    else {
        @0 := const ({impl HasDefaultLen<const N : usize> for Array<bool, const N : usize>}<const N : usize>::LEN)
    }
    storage_dead(@1)
    return
}

// Full name: test_crate::{impl HasDefaultLen<const N : usize> for Array<bool, const N : usize>}::LEN
pub const {impl HasDefaultLen<const N : usize> for Array<bool, const N : usize>}::LEN<const N : usize>: usize = {impl HasDefaultLen<const N : usize> for Array<bool, const N : usize>}::LEN()

// Full name: test_crate::{impl HasDefaultLen<const N : usize> for Array<bool, const N : usize>}
impl<const N : usize> HasDefaultLen<const N : usize> for Array<bool, const N : usize> {
    parent_clause0 = MetaSized<Array<bool, const N : usize>>
    const LEN = {impl HasDefaultLen<const N : usize> for Array<bool, const N : usize>}::LEN<const N : usize>
    non-dyn-compatible
}

// Full name: test_crate::AlsoHasLen
trait AlsoHasLen<Self>
{
    parent_clause0 : [@TraitClause0]: MetaSized<Self>
    parent_clause1 : [@TraitClause1]: HasLen<Self>
    const ALSO_LEN : usize
    non-dyn-compatible
}

// Full name: test_crate::AlsoHasLen::ALSO_LEN
fn ALSO_LEN<Self>() -> usize
where
    [@TraitClause0]: AlsoHasLen<Self>,
{
    let @0: usize; // return
    let @1: usize; // anonymous local

    storage_live(@1)
    @1 := const (@TraitClause0::parent_clause1::LEN) panic.+ const (1 : usize)
    @0 := move (@1)
    return
}

// Full name: test_crate::AlsoHasLen::ALSO_LEN
const ALSO_LEN<Self>: usize
where
    [@TraitClause0]: AlsoHasLen<Self>,
 = ALSO_LEN()

// Full name: test_crate::{impl AlsoHasLen for Array<(), const N : usize>}
impl<const N : usize> AlsoHasLen for Array<(), const N : usize> {
    parent_clause0 = MetaSized<Array<(), const N : usize>>
    parent_clause1 = {impl HasLen for Array<(), const N : usize>}<const N : usize>
    const ALSO_LEN = ALSO_LEN<Array<(), const N : usize>>[{impl AlsoHasLen for Array<(), const N : usize>}<const N : usize>]
    non-dyn-compatible
}



